rfd/0034-clients-rotate-certs.md

```diff
@@ -0,0 +1,187 @@
+
+##### Watch CA rotation state
+
+The agent will use `client.NewWatcher` to watch for updates to the certificate authority's rotation state. The agent will keep track of the current rotation state, and if an event is received where the rotation state is changed to `update_clients`, the agent will refresh the store's certificates and CA certificates.
+
+##### Retrieve new certificates
+
+The agent will use `client.GenerateUserCerts` to retrieve newly signed certificates for its `Store`. The Store's `PublicKey` and username will be provided in order for the Auth server to sign new certificates. The certificate's expiration will also be derived from the agent's `TTL` field.
+
+##### Refresh Client connection
+
+The agent will watch its store's `Refresh` channel in order to refresh the client connection when needed. This may be caused by the Cert Agent writing to the `Store` itself, or by an external actor.
+
+### RBAC restriction
+
+Certificate Agent is intended for automating procedures, not for short-cutting authentication measures already in place. For example, `tsh login` should continue to be the standard authentication method for users logging into the system.
+
+Therefore, new certificates will only be issued if the associated user has the new `reissue_certificates` role option enabled. This option will be false by default and should only be set to true for automation user roles.
+
+Additionally, the CA rotation state can only be watched if the Certificate Agent is allowed perform `read` or `readnosecrets` actions on `cert_authority`.
+
+```yaml
+kind: role
+version: v3
+metadata: 
+  name: cert-agent
+spec:
+  options:
+    reissue_certificates: true
+  allow:
+    rules:
+      - resources: ['cert_authority']
+        verbs: ['readnosecrets']
+```
+
+### Integrations
+
+In addition to being a standalone client, Certificate Agent can be integrated into the API Client, `tctl`, and `tsh` to meet a variety of automation use cases.
+
+#### API Client
+
+The API Client needs a new method to allow the agent to refresh its client's connection.
+
+```go
+func (c *Client) RefreshConnection(creds Credentials) error {
+  // Attempt to connect the client using the updated credentials. 
+  // If it fails, fallback to the former client connection and return the error.
+}
+```
+
+Now, a new client can simply start up a Cert Agent to automatically keep its credentials refreshed. This can be added to the end of the client constructor.
+
+```go
+// connect client to server
+store, ok := client.creds.(Store)
+if ok && config.RunCertAgent {
+  agent := Agent{
+    Client: client,
+    Store: store,
+  }
+  // agent will automatically refresh the store with updated certificates,
+  // refresh the client connection, and close as soon as the client is closed.
+  go agent.Run()
+}
+```
+
+#### tsh and tctl
```

Comment:
@Joerger what do you think about the idea of a separate bot focused on rotation, similarly to `certbot`? It will help you to polish the user experience and the feature set and we will end up with a very useful tool.

For example, we can write a simple bot that will join the cluster and auto renew certificates for some services:

```bash
# Starts a bot, joins the cluster, and creates a certificate in tls and openssh formats:
# reloads 
tbot start --token=token --auth-server=proxy:124 \ # exactly the same syntax as teleport start
        --cert="tls,/var/lib/nginx,nginx -s reload \ # one-liner tuple (format, directory, reload command)
        --cert="openssh,/etc/ssh/,systemctl reload sshd" # restart sshd when certs and CA are updated on rotation or renewal
```

Reply:
I'm more confident in implementing the Cert Agent in the form described by the RFD, which is more similar to Vault autoauth than certbot. Implementing it as a client rather than a standalone service makes it easier to extend to different use cases, such as rotating certificates provided for client authentication, rotating certs created through existing tsh/tctl commands, or implementing Cert Agent as a stand alone service similar to the certbot you've described.

But I'm not opposed to building a teleport-certbot first if you think it will bring out better UX/features, and I wouldn't be surprised if I'm missing some important aspects. What do you see as the main issues with the current Cert Agent design that need improvement, or the main advantages that come with your `tbot` idea?

Reply:
Sorry you can ignore my last reply, I didn't realize your comment was directly relating to the `tsh` and `tctl` idea. I fully agree that implementing it as a standalone service is a better approach/first step. I'll update the RFD with something akin to `tbot`.